File: backend/src/message.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
import { createHash } from 'node:crypto';

import { encode as dagCborEncode } from '@ipld/dag-cbor';
import * as ed from '@noble/ed25519';

/** Schema version of the signed content. Mirrors core's kHearthMessageVersion. */
export const MESSAGE_VERSION = 1;

/** Multihash prefix for sha2-256: code 0x12, length 0x20. */
const SHA256_MULTIHASH_PREFIX = Uint8Array.of(0x12, 0x20);

/** The signed (unsigned-by-id/sig) fields of a message. */
export interface MessageFields {
  version: number;
  author: Uint8Array; // 32-byte Ed25519 public key
  channel: string;
  prev: Uint8Array[]; // parent ids
  timestampMs: number;
  payload: Uint8Array;
}

/** The JSON wire envelope (matches core's `Message.toJson()`, base64url fields). */
export interface WireMessage {
  v: number;
  author: string;
  channel: string;
  prev: string[];
  timestamp: number;
  payload: string;
  sig: string;
  id: string;
  // Multi-device: when present, `sig` is by this device subkey (not the author
  // root), and `cert` is the root's authorisation of that device. The relay
  // verifies the cert chain (so `author` stays authenticated); the receiving
  // client additionally enforces revocation.
  device?: string;
  cert?: WireDeviceCert;
}

/** A device cert on the wire (matches core's `DeviceCert.toJson`). */
export interface WireDeviceCert {
  root: string;
  device: string;
  name: string;
  issued: number;
  sig: string;
}

/**
 * Canonical signed bytes of a device cert — must be byte-for-byte identical to
 * core's `DeviceCert._signedBytes` (canonical CBOR, keys t,name,root,device,
 * issued). @ipld/dag-cbor sorts keys canonically, matching the hand-rolled Dart
 * encoder, same as {@link signedBytes} for messages.
 */
function deviceCertSignedBytes(
  root: Uint8Array,
  device: Uint8Array,
  name: string,
  issued: number,
): Uint8Array {
  return dagCborEncode({
    t: 'hearth/device-cert/v1',
    name,
    root,
    device,
    issued,
  });
}

/**
 * Verifies a device cert binds [device] to [author] (== cert.root) and is
 * signed by that root. Returns false on any mismatch or bad signature.
 */
export async function verifyDeviceCert(
  cert: WireDeviceCert,
  author: Uint8Array,
  device: Uint8Array,
): Promise<boolean> {
  const root = b64urlToBytes(cert.root);
  const certDevice = b64urlToBytes(cert.device);
  if (!bytesEqual(root, author)) return false;
  if (!bytesEqual(certDevice, device)) return false;
  const bytes = deviceCertSignedBytes(root, certDevice, cert.name, cert.issued);
  return ed.verifyAsync(b64urlToBytes(cert.sig), bytes, root);
}

/**
 * The canonical bytes that are signed and hashed — must be byte-for-byte
 * identical to core's `Message.signedBytes()`: canonical dag-cbor with map keys
 * in length-then-bytewise order (v, prev, author, channel, payload, timestamp).
 *
 * @ipld/dag-cbor emits canonical dag-cbor (it sorts keys and uses minimal,
 * definite-length encodings), so it matches the hand-rolled Dart encoder. The
 * interop fixture test guards that this stays true.
 */
export function signedBytes(f: MessageFields): Uint8Array {
  return dagCborEncode({
    v: f.version,
    prev: f.prev,
    author: f.author,
    channel: f.channel,
    payload: f.payload,
    timestamp: f.timestampMs,
  });
}

/** `multihash(sha256(content))` — 34 bytes, matching core's `Message.id`. */
export function computeId(content: Uint8Array): Uint8Array {
  const digest = createHash('sha256').update(content).digest();
  return new Uint8Array([...SHA256_MULTIHASH_PREFIX, ...digest]);
}

/** Verifies an Ed25519 signature over the canonical signed bytes. */
export function verifySignature(
  content: Uint8Array,
  signature: Uint8Array,
  publicKey: Uint8Array,
): Promise<boolean> {
  return ed.verifyAsync(signature, content, publicKey);
}

/**
 * Validates a wire message for the relay: recomputes the canonical id (rejecting
 * a forged id) and checks the signature chain. For a device-signed message the
 * cert must bind the device to `author` and the device must have signed the
 * content — so `author` stays authenticated (rate-limiting keys on it). The
 * client additionally enforces device revocation.
 */
export async function verifyWire(w: WireMessage): Promise<boolean> {
  const fields: MessageFields = {
    version: w.v,
    author: b64urlToBytes(w.author),
    channel: w.channel,
    prev: w.prev.map(b64urlToBytes),
    timestampMs: w.timestamp,
    payload: b64urlToBytes(w.payload),
  };
  const content = signedBytes(fields);
  if (!bytesEqual(computeId(content), b64urlToBytes(w.id))) return false;
  if (w.device) {
    const device = b64urlToBytes(w.device);
    if (!w.cert) return false; // device-signed but no authorisation
    if (!(await verifyDeviceCert(w.cert, fields.author, device))) return false;
    return verifySignature(content, b64urlToBytes(w.sig), device);
  }
  return verifySignature(content, b64urlToBytes(w.sig), fields.author);
}

export function b64urlToBytes(s: string): Uint8Array {
  return new Uint8Array(Buffer.from(s, 'base64url'));
}

export function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i]! ^ b[i]!;
  return diff === 0;
}

export function bytesToHex(b: Uint8Array): string {
  let out = '';
  for (const x of b) out += x.toString(16).padStart(2, '0');
  return out;
}

export function hexToBytes(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

File: core/lib/src/message.dart
// SPDX-License-Identifier: AGPL-3.0-or-later
import 'dart:convert';
import 'dart:typed_data';

import 'package:convert/convert.dart';

import 'codec.dart';
import 'device.dart';
import 'identity.dart';

/// Schema version of the signed content. Bump if the signed fields change.
const int kHearthMessageVersion = 1;

/// Multihash prefix for sha2-256 (code 0x12, length 0x20). Prefixing the digest
/// makes ids self-describing, so the hash algorithm can be migrated later
/// without breaking existing ids (hash agility).
const List<int> _sha256MultihashPrefix = [0x12, 0x20];

/// A signed, content-addressed node in the message DAG.
///
/// [prev] points at the message ids this one "saw", giving causal order without
/// a central clock. [id] is `multihash(sha256(signedBytes))` and [signature] is
/// the author's Ed25519 signature over those same [signedBytes]. Tampering with
/// any signed field invalidates both.
class Message {
  final int version;

  /// 32-byte Ed25519 public key of the author (== their identity).
  final Uint8List author;
  final String channel;

  /// Ids of the messages this one causally follows (its DAG parents/heads).
  final List<Uint8List> prev;

  /// Unix epoch milliseconds. Advisory only — used solely to break ordering
  /// ties between causally-concurrent messages.
  final int timestampMs;

  /// Opaque content. UTF-8 text for now; encrypted bytes later.
  final Uint8List payload;

  /// 64-byte Ed25519 signature over [signedBytes] — by [device] when set, else
  /// by [author] directly.
  final Uint8List signature;

  /// `multihash(sha256(signedBytes))` — 34 bytes.
  final Uint8List id;

  /// The **device subkey** that signed this message, or null when [author]
  /// (the root identity) signed directly. Multi-device: a message is *authored*
  /// by the root but *signed* by one of its devices.
  final Uint8List? device;

  /// The [DeviceCert] proving [device] is authorised by [author]. Non-null iff
  /// [device] is non-null.
  final DeviceCert? cert;

  Message._({
    required this.version,
    required this.author,
    required this.channel,
    required this.prev,
    required this.timestampMs,
    required this.payload,
    required this.signature,
    required this.id,
    this.device,
    this.cert,
  });

  /// The canonical bytes that are signed and hashed — **excludes** [signature]
  /// and [id], which are derived from them.
  ///
  /// Encoded as deterministic (dag-cbor-style) CBOR with map keys in canonical
  /// length-then-bytewise order: `v, prev, author, channel, payload, timestamp`.
  /// This is the cross-language contract: the TypeScript backend must produce
  /// the exact same bytes to verify a signature.
  Uint8List signedBytes() {
    final w = CanonicalCbor()
      ..mapHeader(6)
      ..text('v')
      ..uint(version)
      ..text('prev')
      ..arrayHeader(prev.length);
    for (final p in prev) {
      w.bytes(p);
    }
    w
      ..text('author')
      ..bytes(author)
      ..text('channel')
      ..text(channel)
      ..text('payload')
      ..bytes(payload)
      ..text('timestamp')
      ..uint(timestampMs);
    return w.takeBytes();
  }

  /// Builds, signs, and content-addresses a new message authored by [author]
  /// (the root identity — its pubkey is the author id).
  ///
  /// By default [author] signs directly. For multi-device, pass [signingDevice]
  /// (a device subkey) and its [deviceCert]: the message is still *authored* by
  /// [author] but *signed* by the device, and carries the cert so peers can
  /// verify the root→device→message chain. `device`/`cert` are envelope fields,
  /// not part of [signedBytes], so the content id is identical either way.
  static Future<Message> create({
    required Identity author,
    required String channel,
    required Uint8List payload,
    List<Uint8List> prev = const [],
    int? timestampMs,
    Identity? signingDevice,
    DeviceCert? deviceCert,
  }) async {
    assert(
      signingDevice == null || deviceCert != null,
      'device-signed messages must carry the device cert',
    );
    final ts = timestampMs ?? DateTime.now().toUtc().millisecondsSinceEpoch;
    final signer = signingDevice ?? author;
    final device = signingDevice?.publicKey;
    final fields = Message._(
      version: kHearthMessageVersion,
      author: author.publicKey,
      channel: channel,
      prev: List<Uint8List>.unmodifiable(prev),
      timestampMs: ts,
      payload: payload,
      signature: Uint8List(0), // filled below
      id: Uint8List(0),
    );
    final content = fields.signedBytes();
    final signature = await signer.sign(content);
    final id = await _idFor(content);
    return Message._(
      version: kHearthMessageVersion,
      author: author.publicKey,
      channel: channel,
      prev: List<Uint8List>.unmodifiable(prev),
      timestampMs: ts,
      payload: payload,
      signature: signature,
      id: id,
      device: device,
      cert: deviceCert,
    );
  }

  /// True iff [id] matches the content hash and the signature chain is valid:
  /// either [author] signed directly (classic), or a [device] signed it and a
  /// valid [cert] proves that device is authorised by [author].
  ///
  /// Does **not** check device revocation — that's per-user state the sync/app
  /// layer enforces (drop messages from a device you've revoked).
  Future<bool> verify() async {
    final content = signedBytes();
    final expectedId = await _idFor(content);
    if (!_constTimeEquals(expectedId, id)) return false;
    final dev = device;
    if (dev == null) {
      // Classic: the root identity signed directly.
      return Identity.verifySignature(
        content,
        signature: signature,
        publicKey: author,
      );
    }
    // Device-signed: the cert must bind this device to this author, and the
    // device must have signed the content.
    final c = cert;
    if (c == null) return false;
    if (!_constTimeEquals(c.rootKey, author)) return false;
    if (!_constTimeEquals(c.deviceKey, dev)) return false;
    if (!await c.verify()) return false;
    return Identity.verifySignature(
      content,
      signature: signature,
      publicKey: dev,
    );
  }

  static Future<Uint8List> _idFor(List<int> content) async {
    final digest = await sha256Digest(content);
    return Uint8List.fromList([..._sha256MultihashPrefix, ...digest]);
  }

  /// The id as a hex string, for logging / display.
  String get idHex => hex.encode(id);

  // --- wire envelope (Phase 1: JSON; this is *not* the signed form) ---
  //
  // The receiver reconstructs `signedBytes()` from these fields and re-verifies,
  // so the envelope format itself isn't security-critical. base64url keeps the
  // binary fields JSON-safe.

  Map<String, Object?> toJson() => {
    'v': version,
    'author': base64Url.encode(author),
    'channel': channel,
    'prev': prev.map(base64Url.encode).toList(),
    'timestamp': timestampMs,
    'payload': base64Url.encode(payload),
    'sig': base64Url.encode(signature),
    'id': base64Url.encode(id),
    if (device != null) 'device': base64Url.encode(device!),
    if (cert != null) 'cert': cert!.toJson(),
  };

  static Message fromJson(Map<String, Object?> j) => Message._(
    version: j['v']! as int,
    author: base64Url.decode(j['author']! as String),
    channel: j['channel']! as String,
    prev: (j['prev']! as List<dynamic>)
        .cast<String>()
        .map(base64Url.decode)
        .toList(growable: false),
    timestampMs: j['timestamp']! as int,
    payload: base64Url.decode(j['payload']! as String),
    signature: base64Url.decode(j['sig']! as String),
    id: base64Url.decode(j['id']! as String),
    device: j['device'] == null
        ? null
        : base64Url.decode(j['device']! as String),
    cert: j['cert'] == null
        ? null
        : DeviceCert.fromJson((j['cert']! as Map).cast<String, Object?>()),
  );
}

/// Length-then-content comparison that doesn't short-circuit on content, so id
/// comparison doesn't leak via timing.
bool _constTimeEquals(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  var diff = 0;
  for (var i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff == 0;
}
